File: challenges/programming/esolang/interpreters/py/ctf/player.py
from .enums import Direction

from typing import List, Tuple

class Backpack:
    __stack: List[int]
    
    def __init__(self, stack: List[int] = None):
        if stack == None:
            self.__stack = []
        else:
            self.__stack = stack

    def push(self, val: int):
        self.__stack.append(val)

    def pop(self) -> int:
        return self.__stack.pop()

    def peek(self, depth: int = 0) -> int:
        return self.__stack[-depth - 1]

    def empty(self) -> bool:
        return len(self.__stack) == 0

    def size(self) -> int:
        return len(self.__stack)

    def __str__(self) -> str:
        return str([bytes([c]) for c in list(reversed(self.__stack))])

class Player:
    __position: Tuple[int, int]
    direction: Direction
    __backpack: Backpack
    alive: bool

    def __init__(self, position: Tuple[int, int]):
        self.__position = position
        self.direction = Direction.RIGHT
        self.__backpack = Backpack()
        self.alive = True

    def step(self) -> bool:
        if not self.alive: return False

        if self.direction.value < 2:
            self.__position = (self.__position[0] + self.direction.value * 2 - 1, self.__position[1])
        else:
            self.__position = (self.__position[0], self.__position[1] + (self.direction.value - 2) * 2 - 1)

        return True

    @property
    def position(self) -> Tuple[int, int]:
        return self.__position

    @property
    def backpack(self) -> Backpack:
        return self.__backpack

    def __str__(self) -> str:
        return f"Player {self.__position}: {self.__backpack}"

File: challenges/programming/esolang/interpreters/js/src/player.ts
import { Position, Direction, TileType } from './utils';

export class Backpack {
    private stack: TileType[];

    constructor(stack: TileType[] = []) {
        this.stack = [...stack];
    }

    public clone(): Backpack {
        return new Backpack(this.stack);
    }

    public push(tile: TileType) {
        this.stack.push(tile);
    }

    public pop(): TileType {
        let tile = this.stack.pop();

        if (tile == undefined) throw "Empty Stack";

        return tile;
    }

    public peek(depth: number = 0): TileType {
        return this.stack[this.stack.length - depth - 1];
    }

    public empty(): boolean {
        return this.stack.length == 0;
    }

    public size(): number {
        return this.stack.length;
    }

    public raw(): TileType[] {
        return [...this.stack];
    }
}

export class Player {
    public direction: Direction;
    public alive: boolean;

    private __position: Position;
    private __backpack: Backpack;
    
    constructor(position: Position) {
        this.__position = position.clone();
        this.__backpack = new Backpack();
        this.direction = Direction.RIGHT;
        this.alive = true;
    }

    public get position(): Position {
        return this.__position.clone();
    }

    public get backpack(): Backpack {
        return this.__backpack;
    }

    public raw() {
        return {
            position: this.__position.raw(),
            direction: this.direction,
            stack: this.__backpack.raw()
        }
    }

    public clone(): Player {
        let newPlayer = new Player(this.position);
        newPlayer.__backpack = this.__backpack.clone();
        newPlayer.direction = this.direction;
        newPlayer.alive = this.alive;

        return newPlayer;
    }

    public step(): boolean {
        if(!this.alive) return false;

        switch(this.direction) {
            case Direction.LEFT:
            case Direction.RIGHT:
                this.__position.x += this.direction * 2 - 1;
                break;
            case Direction.UP:
            case Direction.DOWN:
                this.__position.y += (this.direction - 2) * 2 - 1;
                break;
        }

        return true;
    }
}
